packages/flutter/test/cupertino/nav_bar_test.dart

```diff
@@ -2087,6 +2087,212 @@ void main() {
+            slivers: const <Widget>[
+              CupertinoSliverNavigationBar(
+                largeTitle: Text('Large title'),
+                middle: Text('middle'),
+                alwaysShowMiddle: false,
+              ),
+              SliverFillRemaining(
+                child: SizedBox(
+                  height: 1000.0,
+                ),
+              ),
+            ],
+          ),
+        ),
+      ),
+    );
+
+    final RenderAnimatedOpacity? renderOpacity = tester.element(find.text('middle')).findAncestorRenderObjectOfType<RenderAnimatedOpacity>();
+
+    expect(renderOpacity?.opacity.value, 0.0);
+    expect(scrollController.offset, 0.0);
+
+    // Scroll to the halfway point.
+    final TestGesture scrollGesture = await tester.startGesture(tester.getCenter(find.byType(Scrollable)));
+    await scrollGesture.moveBy(const Offset(0.0, -(largeTitleHeight / 2)));
+    await scrollGesture.up();
+    await tester.pumpAndSettle();
+
+    // Expect the large title to snap back to its extended height.
+    expect(scrollController.position.pixels, 0.0);
+    expect(renderOpacity?.opacity.value, 0.0);
+  });
+
+  testWidgets('Large title and bottom snap up when partially scrolled over halfway up in automatic mode', (WidgetTester tester) async {
+    final ScrollController scrollController = ScrollController();
+    const double largeTitleHeight = 52.0;
+    const double bottomHeight = 100.0;
+
+    await tester.pumpWidget(
+      MediaQuery(
+        data: const MediaQueryData(),
+        child: CupertinoApp(
+          home: CustomScrollView(
+            controller: scrollController,
+            slivers: const <Widget>[
+              CupertinoSliverNavigationBar(
+                largeTitle: Text('Large title'),
+                middle: Text('middle'),
+                alwaysShowMiddle: false,
+                bottom: PreferredSize(
+                  preferredSize: Size.fromHeight(bottomHeight),
+                  child: Placeholder(),
+                ),
+                bottomMode: NavigationBarBottomMode.automatic,
+              ),
+              SliverFillRemaining(
+                child: SizedBox(
+                  height: 1000.0,
+                ),
+              ),
+            ],
+          ),
+        ),
+      ),
+    );
+
+    final RenderAnimatedOpacity? renderOpacity = tester.element(find.text('middle')).findAncestorRenderObjectOfType<RenderAnimatedOpacity>();
+
+    expect(renderOpacity?.opacity.value, 0.0);
+    expect(scrollController.offset, 0.0);
+
+    // Scroll to just past the halfway point of the bottom widget.
+    final TestGesture scrollGesture1 = await tester.startGesture(tester.getCenter(find.byType(Scrollable)));
+    await scrollGesture1.moveBy(const Offset(0.0, -(bottomHeight / 2) - 1));
+    await scrollGesture1.up();
+    await tester.pumpAndSettle();
+
+    // Expect the bottom to snap up to the large title.
+    expect(scrollController.position.pixels, bottomHeight);
```

Comment:
Should probably add a check that the bottom widget is hidden directly as well, for completion sake. Instead of just the scroll position.